torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -1937,52 +1937,51 @@ std::vector<Val*> Index::getNonGlobalProducerStridedIndices(
   return strided_inds;
 }
 
-std::vector<Val*> Index::getLinearIndex(
-    TensorView* consumer_tv,
-    const std::vector<kir::ForLoop*>& loops) {
+std::vector<Val*> Index::getRootIndicesWithContiguity(
+    TensorView* tv,
+    const std::vector<kir::ForLoop*>& loops,
+    bool contiguity) {
   // Use domain guard to ignore the contiguity of
   //  consumer tv.
-  TensorDomain* consumer_tv_no_contiguity_domain = nullptr;
-  auto contiguity_vector =
-      std::vector<bool>(consumer_tv->getMaybeRFactorDomain().size(), true);
-  if (consumer_tv->hasRFactor()) {
-    consumer_tv_no_contiguity_domain = IrBuilder::create<TensorDomain>(
-        consumer_tv->getRootDomain(),
-        consumer_tv->getRFactorDomain(),
-        consumer_tv->domain()->domain(),
+  TensorDomain* domain_with_specified_contiguity = nullptr;
+  std::vector<bool> contiguity_vector(
+      tv->getMaybeRFactorDomain().size(), contiguity);
+  if (tv->hasRFactor()) {
+    domain_with_specified_contiguity = IrBuilder::create<TensorDomain>(
+        tv->getRootDomain(),
+        tv->getRFactorDomain(),
+        tv->domain()->domain(),
         contiguity_vector);
   } else {
-    consumer_tv_no_contiguity_domain = IrBuilder::create<TensorDomain>(
-        consumer_tv->getRootDomain(),
-        consumer_tv->domain()->domain(),
-        contiguity_vector);
+    domain_with_specified_contiguity = IrBuilder::create<TensorDomain>(
+        tv->getRootDomain(), tv->domain()->domain(), contiguity_vector);
   }
 
-  ir_utils::TVDomainGuard domain_guard(
-      consumer_tv, consumer_tv_no_contiguity_domain);
+  ir_utils::TVDomainGuard domain_guard(tv, domain_with_specified_contiguity);
 
-  // TODO:
-  //  More optimization on the underlying tensor layout
-  //   will be done in a follow up.
-  return getGlobalConsumerStridedIndices(consumer_tv, loops);
+  IndexFromIdGraph index_from_id_graph = getTensorIndexFromIdGraph(loops, tv);
+  return getRootIndices(tv, loops, index_from_id_graph);
 }
 
-std::vector<Val*> Index::getGlobalConsumerStridedIndices(
-    const TensorView* consumer_tv,
+Val* Index::getLinearIndex(
```

Comment:
My understanding is that `getLinearIndex` would be calculating the indexing assuming all contiguous, while `getPerDimRootIndices` would be assuming all non-contiguous. Is that correct?

Reply:
Would be nice if the naming or comment would reflect that if that's the case. Maybe `getLinearLogicalIndex` and `getPerDimLogicalIndex` ?

Also would be helpful to mention in comment that `Linear` here means linearizing a multi-dimension address into a `linear memory` address, as opposed to mathematical linearity. :)

Reply:
> My understanding is that `getLinearIndex` would be calculating the indexing assuming all contiguous, while `getPerDimRootIndices` would be assuming all non-contiguous. Is that correct?

Yes, correct

> And I think functionality wise `getLinearIndex` should be equivalent to `getPerDimRootIndices` followed by a dot product with strides but  `getLinearIndex`  has a chance of generating better math code due to the contiguity optimization. Just wanted to confirm if that's correct too? 

Yes, also correct.

Reply:
renamed